### PR TITLE
EVA-2779 - Correct assembly in accessioning database for Rattus norvegicus

### PR DESCRIPTION
**Pull request by tcezard**

### CODE REVIEW

File: tasks/eva_2779/correct_assembly_error_in_rat.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+def update_operation_entities(mongo_source, collection_name, id_creation_func, filter_criteria, correction_map, batch_size):
+    collection = mongo_source.mongo_handle[mongo_source.db_name][collection_name]
+    total_updated = 0
+    try:
+        for batch_of_variant in find_documents_in_batch(mongo_source, collection_name, filter_criteria, batch_size):
+            update_statements = []
+            for variant in batch_of_variant:
+                filter_dict = {'_id': variant['_id']}
+                filter_dict.update(filter_criteria)
```
Comment (sundarvenkata-EBI) [inline tasks/eva_2779/correct_assembly_error_in_rat.py:92]: I am wondering why we need an additional criteria when we already have the ID...

File: tasks/eva_2779/correct_assembly_error_in_rat.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        raise e
+    return total_updated
+
+
+def replace_variant_entities(mongo_source, batch_size):
+    # submitted variants on MT chromosome
+    replace_document_with_correct_information(
+        mongo_source, "submittedVariantEntity", get_submitted_SHA1,
+        {'seq': 'GCA_015227675.1', 'contig': 'CM026996.1'},
+        {'seq': 'GCA_015227675.2', 'contig': 'AY172581.1'},
```
Comment (sundarvenkata-EBI) [inline tasks/eva_2779/correct_assembly_error_in_rat.py:122]: Given that we are repeating these values many times in the script, suggest extracting them to a variable...

File: tasks/eva_2779/correct_assembly_error_in_rat.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    replace_document_with_correct_information(
+        mongo_source, "clusteredVariantEntity", get_clustered_SHA1, {'asm': 'GCA_015227675.1'},
+        {'asm': 'GCA_015227675.2'}, batch_size
+    )
+    replace_document_with_correct_information(
+        mongo_source, "dbsnpClusteredVariantEntity", get_clustered_SHA1, {'asm': 'GCA_015227675.1'},
+        {'asm': 'GCA_015227675.2'}, batch_size
+    )
+
+    # MERGED Operations on MT chromosomes
```
Comment (sundarvenkata-EBI) [inline tasks/eva_2779/correct_assembly_error_in_rat.py:166]: Hope you have ensured from the database that the event types laid out here is the exhaustive list..

**Review by sundarvenkata-EBI — APPROVED** [review]
Looks good! I have made a few minor comments. Given the complexity of the script, I suggest running this in tempmongo or a test database first.

**Review by nitin-ebi — APPROVED** [review]

File: tasks/eva_2779/correct_assembly_error_in_rat.py (hunk 1 above)
Comment (tcezard) [reply tasks/eva_2779/correct_assembly_error_in_rat.py:92]: We don't anymore. In a previous version of the script I was using [the $ notation](https://www.mongodb.com/docs/v4.0/reference/operator/update/positional/index.html?_ga=2.68455185.1468278035.1648464773-883181489.1633612269) to update the sub arrays which required that the field replaced was present in the filter. But it turned out to be too complicated.

File: tasks/eva_2779/correct_assembly_error_in_rat.py (hunk 3 above)
Comment (tcezard) [reply tasks/eva_2779/correct_assembly_error_in_rat.py:166]: I check and the database only contains MERGED and RS_SPLIT but the next operation would apply to all types anyway.

**Comment by tcezard** [comment]
> I suggest running this in tempmongo or a test database first.
I would love to test it there but installing python seems complicated. How would you suggest running the code ?

File: tasks/eva_2779/tests/test_correct_assembly_error_in_rat.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    replace_document_with_correct_information, update_operation_entities
+
+
+def calculate_id(rs):
+    h = hashlib.sha1()
+    h.update(str(rs).encode())
+    return h.hexdigest().upper()
+
+
+class TestCorrectAsseblyError(TestCase):
```
Comment (apriltuesday) [inline tasks/eva_2779/tests/test_correct_assembly_error_in_rat.py:16]:
```suggestion
class TestCorrectAssemblyError(TestCase):
```

File: tasks/eva_2779/tests/test_correct_assembly_error_in_rat.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    def test_replace_document_with_correct_information(self):
+        replace_document_with_correct_information(
+            self.mongo_db, "submittedVariantEntity", get_submitted_SHA1,
+            {'seq': 'GCA_015227675.1', 'contig': 'CM026996.1'},
+            {'seq': 'GCA_015227675.2', 'contig': 'AY172581.1'},
+            5
+        )
+        cursor = self.connection_handle[self.accession_db][self.submitted_variants_collection].find()
+        document = cursor.__next__()
```
Comment (apriltuesday) [inline tasks/eva_2779/tests/test_correct_assembly_error_in_rat.py:105]: Can this just be `next()` or do you need the internal method for some reason?

File: tasks/eva_2779/tests/test_correct_assembly_error_in_rat.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            self.mongo_db, 'clusteredVariantOperationEntity', get_clustered_SHA1,
+            {'inactiveObjects.asm': 'GCA_015227675.1', 'eventType': 'MERGED'},
+            {'inactiveObjects.asm': 'GCA_015227675.2',
+             'reason': lambda x: x.replace('GCA_015227675.1', 'GCA_015227675.2')},
+            5
+        )
+        cursor = self.connection_handle[self.accession_db][self.clustered_variants_operation_collection].find()
+        document = cursor.__next__()
+        assert document['reason'] == 'After remapping to GCA_015227675.2, RS IDs mapped to the same locus.'
+        assert document['inactiveObjects'][0]['asm'] == 'GCA_015227675.2'
```
Comment (apriltuesday) [inline tasks/eva_2779/tests/test_correct_assembly_error_in_rat.py:120]: Might want to also check the inactive object's `hashedMessage` is set correctly.

File: tasks/eva_2779/correct_assembly_error_in_rat.py [hunk 7]
```diff
@@ -225,11 +245,19 @@ def main():
...
+    if args.debug:
+        logging_config.add_stdout_handler(logging.DEBUG)
+    else:
+        logging_config.add_stdout_handler()
+
     mongo_source = MongoDatabase(uri=args.mongo_source_uri, secrets_file=args.mongo_source_secrets_file,
                                  db_name="eva_accession_sharded")
-    replace_variant_entities(mongo_source, batch_size=args.batch_size)
+    replace_variant_entities(mongo_source, batch_size=int(args.batch_size))
+    del mongo_source
```
Comment (apriltuesday) [inline tasks/eva_2779/correct_assembly_error_in_rat.py:260]: Why is this `del` necessary?

**Review by apriltuesday — APPROVED** [review]

File: tasks/eva_2779/correct_assembly_error_in_rat.py (hunk 7 above)
Comment (tcezard) [reply tasks/eva_2779/correct_assembly_error_in_rat.py:260]: Don't know !!! 
But the process seems to be taking forever to terminate so I'm making sure that all connection are closed and the `MongoDatabase` closes on `__del__`

**Pull request merged**